packages/amplify-category-function/src/provider-utils/awscloudformation/utils/layerMigrationUtils.ts

```diff
@@ -0,0 +1,191 @@
+import * as path from 'path';
+import { categoryName } from '../../../constants';
+import { layerConfigurationFileName, LegacyFilename, versionHash } from './constants';
+import { loadPluginFromFactory } from './functionPluginLoader';
+import { writeLayerConfigurationFile } from './layerConfiguration';
+import { defaultLayerPermission, LayerPermission, LayerRuntime, PermissionEnum } from './layerParams';
+
+export const enum LegacyPermissionEnum {
+  AwsAccounts = 'awsAccounts',
+  AwsOrg = 'awsOrg',
+  Private = 'private',
+  Public = 'public',
+}
+
+export type LegacyPermission = { type: LegacyPermissionEnum; accounts?: string[]; orgs?: string[] };
+
+type LegacyRuntime = {
+  value: 'nodejs' | 'python';
+  name: 'NodeJS' | 'Python';
+  layerExecutablePath: string;
+  cloudTemplateValue: string;
+};
+
+type LegacyVersionMap = {
+  [key: string]: {
+    permissions: LegacyPermission[];
+    hash?: string;
+  };
+};
+
+const enum LegacyState {
+  NOT_LEGACY,
+  MULTI_ENV_LEGACY,
+  SINGLE_ENV_LEGACY,
+}
+
+const layerVersionMapKey = 'layerVersionMap';
+
+export async function migrateLegacyLayer(context: $TSContext, layerName: string): Promise<boolean> {
+  const layerDirPath = pathManager.getResourceDirectoryPath(undefined, categoryName, layerName);
+  const legacyState = isLegacyLayer(layerName, layerDirPath);
+
+  if (legacyState === LegacyState.NOT_LEGACY) {
+    return false;
+  }
+
+  // TODO add docs link when it's available
+  context.print.warning(
+    '\n⚠️  Amplify updated the way Lambda layers work to better support team workflows and additional features.\n\
+This change requires a migration. Amplify will create a new Lambda layer version even if no layer content changes are made.\n',
+  );
+
+  if (context?.exeInfo?.inputParams?.yes !== true) {
+    const shouldProceedWithMigration = await context.amplify.confirmPrompt('Continue?');
+    if (!shouldProceedWithMigration) {
+      exitOnNextTick(0);
+    }
+  }
+
+  const runtimeCloudTemplateValues: string[] = [];
+  const layerConfiguration: { permissions: LayerPermission[]; runtimes: Partial<LayerRuntime>[]; nonMultiEnv?: boolean } = {
+    permissions: undefined,
+    runtimes: undefined,
+  };
+  let legacyRuntimeArray: LegacyRuntime[];
+  let layerVersionMap: LegacyVersionMap;
+
+  if (legacyState === LegacyState.MULTI_ENV_LEGACY) {
+    legacyRuntimeArray = JSONUtilities.readJson<LegacyRuntime[]>(path.join(layerDirPath, LegacyFilename.layerRuntimes));
+    layerVersionMap = _.get(stateManager.getMeta(), [categoryName, layerName, layerVersionMapKey], {});
+  } else {
+    ({ layerVersionMap, runtimes: legacyRuntimeArray } = JSONUtilities.readJson<{
+      [layerVersionMapKey]: LegacyVersionMap;
+      runtimes: LegacyRuntime[];
+    }>(path.join(layerDirPath, LegacyFilename.layerParameters)));
+    layerConfiguration.nonMultiEnv = true;
+  }
+
+  legacyRuntimeArray
+    .map(legacyRuntime => legacyRuntime.cloudTemplateValue)
+    .forEach((cloudTemplateValue: string) => {
```

Comment:
Can you combine these by using `legacyRuntime.cloudTemplateValue` directly in the `forEach()`. Alternatively, the way the code is currently written, I think this can just be a `map()` call with the result assigned to `runtimeCloudTemplateValues`.